sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -0,0 +1,232 @@
+    assert expand_mul(A.ang_vel_in(N).angle_between(A.x)) == acos(omega/Abs(omega))
+    assert C.masscenter.vel(N).simplify() == - omega*N.z
+    assert C.masscenter.pos_from(P.masscenter) == N.x
+
+    #Both joint pos id defined but different axes
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, parent_joint_pos=N.x, child_joint_pos=A.x, child_axis=A.x+A.y)
+    assert expand_mul(N.x.angle_between(A.x + A.y)) == 0 #Axis are aligned
+    assert (A.x + A.y).express(N).simplify() == sqrt(2)*N.x
+    assert A.dcm(N).simplify() == Matrix([[sqrt(2)/2, -sqrt(2)*cos(theta)/2, -sqrt(2)*sin(theta)/2], \
+        [sqrt(2)/2, sqrt(2)*cos(theta)/2, sqrt(2)*sin(theta)/2], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).express(A).simplify() == (omega*A.x + omega*A.y)/sqrt(2)
+    assert A.ang_vel_in(N).magnitude() == Abs(omega)
+    assert expand_mul(A.ang_vel_in(N).angle_between(A.x + A.y)) == acos(omega/Abs(omega))
+    assert C.masscenter.vel(N).simplify() == (omega * A.z)/sqrt(2)
+    assert C.masscenter.pos_from(P.masscenter) == N.x - A.x
+    assert C.masscenter.pos_from(P.masscenter).express(N).simplify() == \
+        (1 - sqrt(2)/2)*N.x + sqrt(2)*cos(theta)/2*N.y + sqrt(2)*sin(theta)/2*N.z
+    assert C.masscenter.vel(N).express(N).simplify() == - sqrt(2)*omega*sin(theta)/2*N.y + sqrt(2)*omega*cos(theta)/2*N.z
+    assert C.masscenter.vel(N).angle_between(A.x) == pi/2
+
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, parent_joint_pos=N.x, child_joint_pos=A.x, child_axis=A.x+A.y-A.z)
+    assert expand_mul(N.x.angle_between(A.x + A.y - A.z)) == 0 #Axis are aligned
+    assert (A.x + A.y - A.z).express(N).simplify() == sqrt(3)*N.x
+    assert A.dcm(N).simplify() == Matrix([[sqrt(3)/3, -sqrt(6)*sin(theta + pi/4)/3, sqrt(6)*cos(theta + pi/4)/3], \
+        [sqrt(3)/3, sqrt(6)*cos(theta + pi/12)/3, sqrt(6)*sin(theta + pi/12)/3], \
+        [-sqrt(3)/3, sqrt(6)*cos(theta + 5*pi/12)/3, sqrt(6)*sin(theta + 5*pi/12)/3]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).express(A).simplify() == (omega*A.x + omega*A.y - omega*A.z)/sqrt(3)
+    assert A.ang_vel_in(N).magnitude() == Abs(omega)
+    assert expand_mul(A.ang_vel_in(N).angle_between(A.x + A.y-A.z)) == acos(omega/Abs(omega))
+    assert C.masscenter.vel(N).simplify() == (omega * A.y + omega * A.z)/sqrt(3)
+    assert C.masscenter.pos_from(P.masscenter) == N.x - A.x
+    assert C.masscenter.pos_from(P.masscenter).express(N).simplify() == \
+        (1 - sqrt(3)/3)*N.x + sqrt(6)*sin(theta + pi/4)/3*N.y - sqrt(6)*cos(theta + pi/4)/3*N.z
+    assert C.masscenter.vel(N).express(N).simplify() == sqrt(6)*omega*cos(theta + pi/4)/3*N.y + sqrt(6)*omega*sin(theta + pi/4)/3*N.z
+    assert C.masscenter.vel(N).angle_between(A.x) == pi/2
+
+    N, A, P, C = generate_body()
+    m, n = symbols('m n')
+    PinJoint('J', P, C, parent_joint_pos=m*N.x, child_joint_pos=n*A.x, child_axis=A.x+A.y-A.z, parent_axis=N.x-N.y+N.z)
```

Comment:
Figure for the orientation for this one:
![image](https://user-images.githubusercontent.com/276007/123790353-81fb4700-d8de-11eb-91f1-ffe01f9c3bb3.png)

Code:

```python
import sympy as sm
import sympy.physics.mechanics as me

N = me.ReferenceFrame('N')
A = me.ReferenceFrame('A')
As = me.ReferenceFrame('As')
Ns = me.ReferenceFrame('Ns')

a1, a2, a3 = sm.symbols('a1, a2, a3')
n1, n2, n3 = sm.symbols('n1, n2, n3')

theta = me.dynamicsymbols('theta')
repl = {a1: 1, a2: 1, a3: -1, n1: 1, n2: -1, n3: 1}

rP = n1*N.x + n2*N.y + n3*N.z
uP = me.cross(N.x, rP).normalize()
thetaP = sm.acos(me.dot(N.x, rP.normalize()))

rC = a1*A.x + a2*A.y + a3*A.z
uC = me.cross(A.x, rC).normalize()
thetaC = sm.acos(me.dot(A.x, rC.normalize()))

u = uC.dot(A.x)*As.x + uC.dot(A.y)*As.y + uC.dot(A.z)*As.z

Ns.orient_axis(N, uP, thetaP)
As.orient_axis(Ns, Ns.x, theta)
A.orient_axis(As, -u, thetaC)

print(sm.trigsimp(A.dcm(N).xreplace(repl)))
print('\n')
print(A.ang_vel_in(N).express(N).xreplace(repl).simplify())
print('\n')
print(A.ang_vel_in(N).express(A).xreplace(repl).simplify())
```

Output:

```
Matrix([[1/3 - 2*cos(theta(t))/3, -2*sin(theta(t) + pi/6)/3 - 1/3, 2*cos(theta(t) + pi/3)/3 + 1/3], [2*cos(theta(t) + pi/3)/3 + 1/3, 2*cos(theta(t))/3 - 1/3, 2*sin(theta(t) + pi/6)/3 + 1/3], [-2*sin(theta(t) + pi/6)/3 - 1/3, 2*cos(theta(t) + pi/3)/3 + 1/3, 2*cos(theta(t))/3 - 1/3]])


sqrt(3)*Derivative(theta(t), t)/3*N.x - sqrt(3)*Derivative(theta(t), t)/3*N.y + sqrt(3)*Derivative(theta(t), t)/3*N.z


sqrt(3)*Derivative(theta(t), t)/3*A.x + sqrt(3)*Derivative(theta(t), t)/3*A.y - sqrt(3)*Derivative(theta(t), t)/3*A.z
```

Reply:
These match your results for the orientation and angular velocity.

Reply:
> These match your results for the orientation and angular velocity.

And DCM too